README.md

```diff
@@ -24,52 +24,86 @@ The 2048 Game project is a Python implementation of the popular 2048 puzzle game
 - **Past Grid Tracking**: Keep track of the previous grid states using a stack.
 - **Score Tracking**: Each move the score get updated
-- **Game Over and Victory Detection**: Detect when the game is over or when the player wins.
+- **Undo Function**: Allows players to revert to the previous grid state, facilitating strategic moves and error correction.
+- **Game Over and Victory Detection**: Detect when the game is over or when the player wins. And also **Quit** possibility
+- **Ranking System**: Maintaining a ranking of 10 most high scores.
+- **Tree-Based Move Suggestions**: Provides approximate suggestions for optimal moves using a tree structure.
 
 ## Data Structures Used
 
 - **2D Array**: Representing the game grid.
 - **Stack** (linked list-based): Storing past grid states.
 - **Dictionary**: Managing key game features.
 - **Set**: Use a set to keep track of empty celleces (tiles with value 0)
+- **Tree**: Used in the tree-based move suggestion system.
 
 ## Algorithms
 
 - **Merge Algorithm**: Sum two cells adjacent in the same direction of movement.
 - **Random Number Generator Algorithm**: Generate random numbers in the grid.
 - **Move Algorithm**: Move the grid in the specified direction.
+- **Sorting Algorithms (QuickSort)**: Used in ranking system to sort the high scores.
 
 ## Code Structure Explanation
 
-The code is organized into several components to enhance readability and maintainability.
+The code is organized into several components to enhance readability and maintainability:
+
+- **`main.py`**: The entry point of the game. It initializes the `GameBoard` and runs the main game loop.
+
+- **`src` folder**: Contains various modules and classes:
+  
+  - _game_class.py_: Contains the GameBoard class with methods for game mechanics.
+
+  - _stack.py_: Implements the Stack class for undo functionality.
+
+  - _tree_class.py_: Defines the TreePossibilities class for move suggestions.
+
+  - _tree.py:_ Implements a tree with an array of children
+
+
 
 ### `GameBoard` Class
 
 The `GameBoard` class is the core of the game logic and user interaction. It contains the following methods:
 
-- **`__init__(self, size)`**: Initializes the game board with a specified size and sets up the initial grid with empty cells.
+- **`__init__(self, size)`**: Initializes the game board with a specified size ( 4 x 4 ) and sets up the initial grid with empty cells.
 
 - **`initialize_empty_coordinates`**: Go troght all the grid as save the empty spaces in a set()
+   - Time Complexity: O(n^2) - where n is the size of the grid. This method iterates through each cell of the grid.
+
 
 - **`update(self)`**: Prints the current state of the game grid to the terminal.
+  - Time Complexity: O(n^2) - as it iterates through each cell for printing.
+
+- **`read_user_input(self)`**: Listens for keyboard inputs using the `keyboard` library to determine the user's desired move direction, or actions.
 
```

Comment:
describe fast what all the input can do:

        direction_keys = {'left': 'left arrow', 'right': 'right arrow', 'up': 'up arrow', 'down': 'down arrow', 'undo': 'u', 'ranking': 'r', 'quit': 'q'}